wazo_calld/plugins/voicemails/resources.py

```diff
@@ -254,3 +264,107 @@ def _validate_message_id(message_id):
+    def get(self, voicemail_id, greeting):
+        voicemail_id = _validate_voicemail_id(voicemail_id)
+        greeting = _validate_greeting(greeting)
+        return self._get(voicemail_id, greeting)
+
+    @required_acl('calld.voicemails.{voicemail_id}.greetings.{greeting}.update')
+    def put(self, voicemail_id, greeting):
+        voicemail_id = _validate_voicemail_id(voicemail_id)
+        greeting = _validate_greeting(greeting)
+        return self._put(voicemail_id, greeting)
+
+    @required_acl('calld.voicemails.{voicemail_id}.greetings.{greeting}.delete')
+    def delete(self, voicemail_id, greeting):
+        voicemail_id = _validate_voicemail_id(voicemail_id)
+        greeting = _validate_greeting(greeting)
+        return self._delete(voicemail_id, greeting)
+
+
+class UserVoicemailGreetingResource(_BaseVoicemailGreetingResource):
+
+    def __init__(self, auth_client, voicemails_service):
+        super().__init__(voicemails_service)
+        self._auth_client = auth_client
+
+    @required_acl('calld.users.me.voicemails.greetings.{greeting}.read')
+    def get(self, greeting):
+        voicemail_id = _get_voicemail_id_from_request(self._auth_client, self._voicemails_service)
+        greeting = _validate_greeting(greeting)
+        return self._get(voicemail_id, greeting)
+
+    @required_acl('calld.users.me.voicemails.greetings.{greeting}.update')
+    def put(self, greeting):
+        voicemail_id = _get_voicemail_id_from_request(self._auth_client, self._voicemails_service)
+        greeting = _validate_greeting(greeting)
+        return self._put(voicemail_id, greeting)
+
+    @required_acl('calld.users.me.voicemails.greetings.{greeting}.delete')
+    def delete(self, greeting):
+        voicemail_id = _get_voicemail_id_from_request(self._auth_client, self._voicemails_service)
+        greeting = _validate_greeting(greeting)
+        return self._delete(voicemail_id, greeting)
+
+
+class VoicemailGreetingCopyResource(_BaseVoicemailGreetingResource):
+
+    @required_acl('calld.voicemails.{voicemail_id}.greetings.{greeting}.copy.create')
+    def post(self, voicemail_id, greeting):
+        voicemail_id = _validate_voicemail_id(voicemail_id)
+        greeting = _validate_greeting(greeting)
+        return self._copy(voicemail_id, greeting)
+
+
+class UserVoicemailGreetingCopyResource(_BaseVoicemailGreetingResource):
+
+    def __init__(self, auth_client, voicemails_service):
+        super().__init__(voicemails_service)
+        self._auth_client = auth_client
+
+    @required_acl('calld.users.me.voicemails.greetings.{greeting}.copy.create')
+    def post(self, greeting):
+        voicemail_id = _get_voicemail_id_from_request(self._auth_client, self._voicemails_service)
+        greeting = _validate_greeting(greeting)
+        return self._copy(voicemail_id, greeting)
+
+
+def _validate_greeting(greeting):
+    if greeting in VALID_GREETINGS:
+        return greeting
+    raise InvalidVoicemailGreeting(greeting)
```

Comment:
Use marshmallow and try-except to reraise the good exception OR raise the exception from a custom marshmallow validator (with `@validate(<field>)` decorator).
If it not necessary, do not make field validation outside schema

Reply:
I have copying the scheme of the rest of the file, and wondering why marshmallow was not used. I will fix all of them.

Reply:
Oh you right, it is the `VoicemailGreetingCopy` without validation for greeting that confused me. But you have fixed it :+1: